clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -237,6 +238,20 @@ public boolean completeExceptionally(
         return done(ProduceResponse.INVALID_OFFSET, RecordBatch.NO_TIMESTAMP, topLevelException, recordExceptions);
     }
 
+    /**
+     * Get all record futures for this batch.
+     * This is used by flush() to wait on individual records rather than the batch-level future.
+     * When batches are split, individual futures are chained to the new batches,
+     * ensuring flush() waits for all split batches to complete.
+     *
+     * @return List of FutureRecordMetadata for all records in this batch
+     */
+    public List<FutureRecordMetadata> recordFutures() {
+        return thunks.stream()
+            .map(thunk -> thunk.future)
```

Comment:
The downside of this approach is the potential overhead due to record level checking. Could you do some perf test to compare the flush() time with and w/o this PR with say a few thousands pending records?

Another option is to also chain ProduceRequestResult when splitting the batches. This allows us to check the status at the request level in the flush() call.